extensions/agroal/runtime/src/main/java/io/quarkus/agroal/runtime/AbstractDataSourceProducer.java

```diff
@@ -246,4 +267,40 @@ public void stop() {
             }
         }
     }
+
+    /**
+     * Wrapper to facilitate filtering & sorting of applicable CDI produced AgroalDataSourceListeners
+     */
+    private class PrioritizedDataSourceListenerWrapper implements Comparable<PrioritizedDataSourceListenerWrapper> {
+        final AgroalDataSourceListener listener;
+        final int priority;
+        final String dataSourceName;
+
+        PrioritizedDataSourceListenerWrapper(final AgroalDataSourceListener listener, final int priority,
+                final String dataSourceName) {
+            this.listener = listener;
+            this.priority = priority;
+            this.dataSourceName = dataSourceName;
+        }
+
+        PrioritizedDataSourceListenerWrapper(final Bean<AgroalDataSourceListener> listener) {
+            this.listener = (AgroalDataSourceListener) beanManager.getReference(listener, listener.getBeanClass(),
+                    beanManager.createCreationalContext(listener));
+
+            Priority p = listener.getBeanClass().getAnnotation(Priority.class);
+            this.priority = p != null ? p.value() : Interceptor.Priority.APPLICATION;
+
+            io.quarkus.agroal.DataSource ds = listener.getBeanClass().getAnnotation(io.quarkus.agroal.DataSource.class);
```

Comment:
`io.quarkus.agroal.DataSource` is a qualifier so there's no need to use reflection to obtain the name. `Bean.getQualifiers()` could be used to obtain the annotation instance.

Reply:
When I tried that, all I was getting was DEFAULT and Any

Reply:
> I like the idea but I'd try to avoid using `BeanManager` and `Bean` APIs.

Yeah. I wasn't sure how to go about that without introducing a bunch of a new stuff into Quarkus, and I didn't want to go that route without having some discussion with the folks in charge. :-)

Reply:
You will get `Default` and `Any` for the default DS. For a DS defined like [this](https://github.com/quarkusio/quarkus/blob/master/extensions/agroal/deployment/src/test/resources/application-multiple-datasources.properties#L7-L11), you'll get `Any`, `DataSource("users")` and `Named("users")` (not sure why, maybe @Sanne knows?).

Reply:
Except this isn't getting the DataSource -- this is getting the @DataSource qualifier off the @ApplicationScoped listener bean.

Example of a listener that gets picked up by this: (wow this really was a bad PR, huh?)
```
@DataSource("someDataSourceName")
@ApplicationScoped
@Priority(10)
public class AuditingDataSourceListener implements AgroalDataSourceListener {
...
}
```

In this case, the qualifier is used to determine what datasource to apply the listener to (rather than all of them).